src/algorithms/onnx/InclusiveKinematicsML.cc

```diff
@@ -0,0 +1,127 @@
+    ss << v[v.size() - 1];
+    return ss.str();
+  }
+
+  template <typename T>
+  Ort::Value vec_to_tensor(std::vector<T>& data, const std::vector<std::int64_t>& shape) {
+    Ort::MemoryInfo mem_info =
+        Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault);
+    auto tensor = Ort::Value::CreateTensor<T>(mem_info, data.data(), data.size(), shape.data(), shape.size());
+    return tensor;
+  }
+
+  void InclusiveKinematicsML::init(std::shared_ptr<spdlog::logger>& logger) {
+    m_log = logger;
+
+    // onnxruntime setup
+    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "inclusive-kinematics-ml");
+    Ort::SessionOptions session_options;
+    try {
+      m_session = Ort::Session(env, m_cfg.modelPath.c_str(), session_options);
+
+      // print name/shape of inputs
+      Ort::AllocatorWithDefaultOptions allocator;
+      m_log->debug("Input Node Name/Shape:");
+      for (std::size_t i = 0; i < m_session.GetInputCount(); i++) {
+        m_input_names.emplace_back(m_session.GetInputNameAllocated(i, allocator).get());
+        m_input_shapes.emplace_back(m_session.GetInputTypeInfo(i).GetTensorTypeAndShapeInfo().GetShape());
+        m_log->debug("\t{} : {}", m_input_names.at(i), print_shape(m_input_shapes.at(i)));
+      }
+
+      // print name/shape of outputs
+      m_log->debug("Output Node Name/Shape:");
+      for (std::size_t i = 0; i < m_session.GetOutputCount(); i++) {
+        m_output_names.emplace_back(m_session.GetOutputNameAllocated(i, allocator).get());
+        m_output_shapes.emplace_back(m_session.GetOutputTypeInfo(i).GetTensorTypeAndShapeInfo().GetShape());
+        m_log->debug("\t{} : {}", m_output_names.at(i), print_shape(m_output_shapes.at(i)));
+      }
+
+      // convert names to char*
+      m_input_names_char.resize(m_input_names.size(), nullptr);
+      std::transform(std::begin(m_input_names), std::end(m_input_names), std::begin(m_input_names_char),
+                     [&](const std::string& str) { return str.c_str(); });
+      m_output_names_char.resize(m_output_names.size(), nullptr);
+      std::transform(std::begin(m_output_names), std::end(m_output_names), std::begin(m_output_names_char),
+                     [&](const std::string& str) { return str.c_str(); });
+
+    } catch(std::exception& e) {
+      m_log->error(e.what());
+    }
+  }
+
+  void InclusiveKinematicsML::process(
+      const InclusiveKinematicsML::Input& input,
+      const InclusiveKinematicsML::Output& output) const {
+
+    const auto [electron, da] = input;
+    auto [ml] = output;
+
+    // Require valid inputs
+    if (electron->size() == 0 || da->size() == 0) {
+      m_log->debug("skipping because input collections have no entries");
+      return;
+    }
+
+    // Assume model has 1 input nodes and 1 output node.
+    if (m_input_names.size() != 1 || m_output_names.size() != 1) {
+      m_log->debug("skipping because model has incorrect input and output size");
+      return;
+    }
```

Comment:
Correct. The idea is that we want to verify at least some quantities about the model to make sure it is consistent with the context it is asked to operate on. This will likely be even stricter and involve some variable name introspection which will have to agree with the input tensor names from the model.